File: src/lib/nn-math.ts
// Neural-network math shared by the Dense/Conv1D nodes, the 3D weight-web
// views, and the execution engine. Split out of execution-helpers.ts to keep
// that file under the repo's 500-line module guardrail.

// Deterministic 32-bit PRNG (mulberry32). The Dense Layer derives its weight
// matrix from this so weights stay stable across re-renders/reloads AND match
// the Python engine exactly — _mulberry32 in backend/engine/nn_math.py is a
// bit-for-bit port; change one and you must change both.
export function mulberry32(seed: number): () => number {
  let a = seed >>> 0;
  return () => {
    a = (a + 0x6d2b79f5) >>> 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 15), t | 1) >>> 0;
    t = (((t + (Math.imul(t ^ (t >>> 7), t | 61) >>> 0)) >>> 0) ^ t) >>> 0;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

/**
 * Weight matrix for a Dense Layer: weights[neuron][input] in [-1, 1),
 * generated row-by-row from the seeded PRNG (same order as the Python mirror).
 */
export function generateWeights(seed: number, inputSize: number, neurons: number): number[][] {
  const rand = mulberry32(seed);
  const weights: number[][] = [];
  for (let j = 0; j < neurons; j++) {
    const row: number[] = [];
    for (let i = 0; i < inputSize; i++) {
      row.push(rand() * 2 - 1);
    }
    weights.push(row);
  }
  return weights;
}

// Coerces a port value into a numeric vector (AI Model nodes pass arrays).
export function toNumberVector(v: any): number[] {
  if (!Array.isArray(v)) return [];
  return v.map((x) => {
    const n = Number(x);
    return isNaN(n) ? 0 : n;
  });
}

export const CONV_ACTIVATIONS: Record<string, (z: number) => number> = {
  relu: (z) => Math.max(0, z),
  sigmoid: (z) => 1 / (1 + Math.exp(-z)),
  tanh: (z) => Math.tanh(z),
};

export function conv1dOutputPositions(inputLen: number, kernelSize: number, stride: number): number {
  if (inputLen < kernelSize || kernelSize < 1 || stride < 1) return 0;
  return Math.floor((inputLen - kernelSize) / stride) + 1;
}

/**
 * Runs a real 1D convolution: `filters` small kernels (generated from `seed`,
 * same mulberry32 PRNG as Dense Layer) each slide across `input` with the
 * given stride, producing one activated value per window. Filters' outputs
 * are concatenated into a single feature-map vector — unlike a Dense Layer,
 * each output only ever depends on a local neighborhood of the input.
 */
export function conv1dForward(
  input: number[],
  seed: number,
  kernelSize: number,
  filters: number,
  stride: number,
  activation: string
): number[] {
  const positions = conv1dOutputPositions(input.length, kernelSize, stride);
  if (positions === 0) return [];
  const kernels = generateWeights(seed, kernelSize, filters); // kernels[f][k]
  const fn = CONV_ACTIVATIONS[activation] ?? CONV_ACTIVATIONS.relu;
  const out: number[] = [];
  for (let f = 0; f < filters; f++) {
    for (let p = 0; p < positions; p++) {
      let z = 0;
      for (let k = 0; k < kernelSize; k++) z += kernels[f][k] * input[p * stride + k];
      out.push(fn(Math.max(-60, Math.min(60, z))));
    }
  }
  return out;
}

/**
 * Full [output][input] weight matrix for a Conv1D Layer, mostly zero except
 * each output's local receptive field — lets the 3D viewer and inline weight
 * web reuse the exact same dense-matrix rendering path as a Dense Layer while
 * still only lighting up the real local connections.
 */
export function conv1dFullWeights(
  inputLen: number,
  seed: number,
  kernelSize: number,
  filters: number,
  stride: number
): number[][] {
  const positions = conv1dOutputPositions(inputLen, kernelSize, stride);
  if (positions === 0) return [];
  const kernels = generateWeights(seed, kernelSize, filters);
  const weights: number[][] = [];
  for (let f = 0; f < filters; f++) {
    for (let p = 0; p < positions; p++) {
      const row = new Array(inputLen).fill(0);
      for (let k = 0; k < kernelSize; k++) row[p * stride + k] = kernels[f][k];
      weights.push(row);
    }
  }
  return weights;
}

File: backend/engine/nn_math.py
def _mulberry32(seed: int):
    """Bit-for-bit port of mulberry32 in src/lib/execution-helpers.ts, so a
    Dense Layer's weight web is identical in the live preview and the /run
    engine. Change one and you must change both."""
    a = int(seed) & 0xFFFFFFFF

    def rand() -> float:
        nonlocal a
        a = (a + 0x6D2B79F5) & 0xFFFFFFFF
        t = a
        t = ((t ^ (t >> 15)) * (t | 1)) & 0xFFFFFFFF
        t = (((t + (((t ^ (t >> 7)) * (t | 61)) & 0xFFFFFFFF)) & 0xFFFFFFFF) ^ t) & 0xFFFFFFFF
        return (t ^ (t >> 14)) / 4294967296

    return rand


def _generate_weights(seed: int, input_size: int, neurons: int):
    """weights[neuron][input] in [-1, 1) — same generation order as the TS
    generateWeights so both sides produce the same matrix."""
    rand = _mulberry32(seed)
    return [[rand() * 2 - 1 for _ in range(input_size)] for _ in range(neurons)]


def _conv1d_output_positions(input_len: int, kernel_size: int, stride: int) -> int:
    if input_len < kernel_size or kernel_size < 1 or stride < 1:
        return 0
    return (input_len - kernel_size) // stride + 1


def _conv1d_forward(values, seed, kernel_size, filters, stride, activation):
    """Bit-for-bit port of conv1dForward in src/lib/execution-helpers.ts —
    each filter's kernel slides across `values`, so every output only depends
    on a local window, not the whole input (unlike a Dense Layer)."""
    import math
    positions = _conv1d_output_positions(len(values), kernel_size, stride)
    if positions == 0:
        return []
    kernels = _generate_weights(seed, kernel_size, filters)  # kernels[f][k]
    out = []
    for f in range(filters):
        for p in range(positions):
            z = sum(kernels[f][k] * values[p * stride + k] for k in range(kernel_size))
            z = max(-60.0, min(60.0, z))
            if activation == "sigmoid":
                out.append(1.0 / (1.0 + math.exp(-z)))
            elif activation == "tanh":
                out.append(math.tanh(z))
            else:
                out.append(max(0.0, z))  # relu
    return out


def _to_number_vector(v) -> list:
    if not isinstance(v, list):
        return []
    out = []
    for x in v:
        try:
            out.append(float(x))
        except (TypeError, ValueError):
            out.append(0.0)
    return out
